server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java

```diff
@@ -85,71 +88,59 @@ protected void onShardGroupFailure(int shardIndex, Exception exc) {
     }
 
-    ShardSearchRequest buildShardSearchRequestInternal(SearchShardIterator shardIt) {
-        ShardSearchRequest request = super.buildShardSearchRequest(shardIt);
-        if (resultConsumer == null) {
-            return request;
+    private synchronized void mergeTopDocs(QuerySearchResult result) {
+        if (result.isNull()
+                || hasPrimaryFieldSort == false
+                || result.topDocs().topDocs instanceof TopFieldDocs == false) {
+            return;
         }
-        List<TopFieldDocs> topDocsList = resultConsumer.getRemainingTopDocs().stream()
-            .filter(obj -> obj instanceof TopFieldDocs)
-            .map(obj -> (TopFieldDocs) obj)
-            .collect(Collectors.toList());
-        if (topDocsList.isEmpty()) {
-            return request;
+        // merge the current best bottom field doc with the new query result
+        TopFieldDocs topDocs = (TopFieldDocs) result.topDocs().topDocs;
+        final ScoreDoc[] bottomDocs;
+        if (topDocs.scoreDocs.length == topDocsSize) {
+            FieldDoc bottom = (FieldDoc) topDocs.scoreDocs[topDocsSize - 1];
+            bottomDocs = new FieldDoc[] { new FieldDoc(bottom.doc, bottom.score, bottom.fields, result.getShardIndex()) };
+            bottomDocs[0].shardIndex = result.getShardIndex();
+        } else {
+            bottomDocs = new ScoreDoc[0];
         }
-        SearchSourceBuilder source = request.source();
-        int trackTotalHits = source.trackTotalHitsUpTo() == null ? DEFAULT_TRACK_TOTAL_HITS_UP_TO : source.trackTotalHitsUpTo();
-        long totalHits = 0;
-        FieldDoc bestBottom = null;
-        SortField primarySort = topDocsList.get(0).fields[0];
-        FieldComparator fieldComparator = primarySort.getComparator(1, 0);
-        int reverseMul = primarySort.getReverse() ? -1 : 1;
-        // we don't want to perform a costly merge to find the best bottom field doc so
-        // we just check pick the best bottom document among the available buffer. This
-        // means that we don't have the true-best bottom but this avoids running a partial
-        // merge too often.
-        for (TopFieldDocs topDocs : topDocsList) {
-            totalHits += topDocs.totalHits.value;
-            if (topDocs.scoreDocs.length == topDocsSize) {
-                FieldDoc cand = (FieldDoc) topDocs.scoreDocs[topDocsSize-1];
-                if (bestBottom == null ||
-                        fieldComparator.compareValues(cand.fields[0], bestBottom.fields[0]) * reverseMul < 0) {
-                    bestBottom = cand;
-                }
-            }
+        TopFieldDocs toMerge = new TopFieldDocs(topDocs.totalHits, bottomDocs, topDocs.fields);
+        if (bottomTopDocs == null) {
+            bottomTopDocs = toMerge;
+        } else {
+            final Sort sort = new Sort(bottomTopDocs.fields);
+            bottomTopDocs = TopFieldDocs.merge(sort, 0, 1, new TopFieldDocs[]{bottomTopDocs, toMerge}, false);
```

Comment:
Previously using `fieldComparator.compareValues` looks to be faster than using `TopFieldDocs.merge` (as TopFieldDocs.merge doing a lot of stuff that we don't really need),  especially considering that `mergeTopDocs` is a synchronized function, and we want to be quite fast.